Add goroutine core affinity support for RP2040/RP2350 systems

amken3d opened the pull request:
This PR proposes
- Support for CPU core pinning and affinity for tasks and goroutines.
- Updated the scheduler to respect affinity constraints with separate queues for pinned and shared tasks.
- Added new runtime API functions `LockToCore`, `UnlockFromCore`, `GetAffinity`, and `CurrentCPU`.
- Example program demonstrates core pinning and unpinned execution behavior.

## API Functions
 
### `runtime.NumCPU() int`
Returns the number of CPU cores available (returns 2 on RP2040/RP2350).
 
### `runtime.CurrentCPU() int`
Returns the current CPU core number (0 or 1).
 
### `runtime.LockToCore(core int)`
Pins the current goroutine to the specified core:
- `core = 0` - Pin to core 0
- `core = 1` - Pin to core 1
- `core = -1` - Unpin (allow running on any core)
 
Panics if core is invalid (not -1, 0, or 1).
 
### `runtime.UnlockFromCore()`
Unpins the current goroutine, allowing it to run on any core.
Equivalent to `runtime.LockToCore(-1)`.
 
### `runtime.GetAffinity() int`
Returns the current goroutine's CPU affinity:
- Returns `-1` if not pinned (can run on any core)
- Returns `0` or `1` if pinned to that specific core
 
###  Example program included in the examples directory
- Tested on both pico and pico2

- Output of example program

```
=== Core Pinning Example ===                                             
Number of CPU cores: 2                                                   
Main starting on core: 0                                                 
                                                                         
Main pinned to core: 0

Core 0 (main): 0 on CPU 0
Worker pinned to core: 1
  Core 1 (worker): 0 on CPU 0
Unpinned worker starting, affinity: 0
    Unpinned worker: 0 on CPU 0
Core 0 (main): 1 on CPU 0
    Unpinned worker: 1 on CPU 0
Core 0 (main): 2 on CPU 0
  Core 1 (worker): 2 on CPU 1
    Unpinned worker: 2 on CPU 0                                                 
Core 0 (main): 3 on CPU 0                                                       
  Core 1 (worker): 3 onCPU 1                                                    
Core 0 (main): 4 on CPU 0                                                       
  Core 1 (worker): 4 on CPU 1                                                   
    Unpinned worker: 3 on CPU 0                                                 
Core 0 (main): 5 on CPU 0                                                       
  Core 1 (worker): 5 on CPU 1                                                   
    Unpinned worker: 4 on CPU 0                                                 
Core 0 (main): 6 on CPU 0                                                       
  Core 1 (worker): 6 on CPU 1                                                   
Core 0 (main): 7 on CPU 0                                                       
  Core 1 (worker): 7 on CPU 1                                                   
    Unpinned worker: 5 on CPU 0                                                 
Core 0 (main): 8 on CPU 0                                                       
  Core 1 (worker): 8 on CPU 1                                                   
    Unpinned worker: 6 on CPU 0                                                 
Core 0 (main): 9 on CPU 0                                                       
  Core 1 (worker): 9 on CPU 1                                                   
    Unpinned worker: 7 on CPU 0                                                 
                                                                                
Main unpinned, affinity: -1                                                     
Unpinned main on CPU 0                                                          
  Core 1 worker finished                                                        
Unpinned main on CPU 0                                                          
    Unpinned worker: 8 on CPU 0                                                 
Unpinned main on CPU 0                                                          
    Unpinned worker: 9 on CPU 0                                                 
Unpinned main on CPU 0                                                          
Unpinned main on CPU 1                                                          
    Unpinned worker finished                                                    
                                                                                
Example complte!   
```

amken3d committed:
Add goroutine core affinity support for RP2040/RP2350 systems

- Introduced support for CPU core pinning and affinity for tasks and goroutines.
- Updated the scheduler to respect affinity constraints with separate queues for pinned and shared tasks.
- Added new runtime API functions `LockToCore`, `UnlockFromCore`, `GetAffinity`, and `CurrentCPU`.
- Example program demonstrates core pinning and unpinned execution behavior.

eliasnaur commented:
Do you actually care about the particular core? If not, are the existing `runtime.LockOSThread` and `runtime.UnlockOSThread` calls sufficient to lock/unlock a goroutine to a core?

amken3d commented:
This is what I see for LockOsThread

// LockOSThread wires the calling goroutine to its current operating system thread.
// Stub for now
// Called by go1.18 standard library on windows, see https://github.com/golang/go/issues/49320
func LockOSThread() {
}

// UnlockOSThread undoes an earlier call to LockOSThread.
// Stub for now
func UnlockOSThread() {
}

There seems to be no implementation behind it. 

For the RP2, since it is symmetrical multi processor, it probably doesn't matter which exact core. But for something like StM32h7, it would matter which core you pin to. (I know we don't support multicore on it yet)

eliasnaur commented:
I know. What I'm saying is to change `LockOSThread` to mean "lock the current goroutine to a core" (when using the `cores` scheduler).

amken3d commented:
Fair point. That seems reasonable to me. I can use those function names instead. 
The only issue I see is that Lock OsThread can't take any arguments. I'd like to be able to tell which core to lock to.

eliasnaur commented:
Right. So `LockOSThread` is enough for use cases where you only care about exclusive access to a some core. For heterogeneous cores, I suggest:

- Move the API to package `machine`.
- Drop `CurrentCPU` - it's racy (its result may be invalidated at any time).
- Drop `GetAffinity` - it seems that code that pin itself to a particular core should know what it's doing.
- Drop the `-1` special case from `LockToCore`
- Rename `LockToCore` to `LockCore` to mimic `LockOSThread` naming. Rename `UnlockFromCore` to `UnlockCore` for the same reason.

An important issue to think about is what happens if the requested core is busy? `LockOSThread` doesn't have this problem (*some* core must be running it).

amken3d committed:
Refactored based on Elias' comments

- Dropped CurrentCPU
- Dropped GetAffinity
- Renamed LockToCore to LockCore to mimic LockOSThread naming.
- Updated examples program

amken3d commented:
Looks like it passed all checks except the macos(13) test with this error

This is a scheduled macos-13 brownout. The macOS-13 based runner images are being deprecated. For more details, see https://github.com/actions/runner-images/issues/13046.

deadprogram commented:
@amken3d I just created https://github.com/tinygo-org/tinygo/pull/5093 to address the macOS 13 runner deprecation.

eliasnaur commented on src/runtime/runtime.go:
```diff
@@ -98,14 +98,18 @@ func os_sigpipe() {
-// Stub for now
+// On microcontrollers with multiple cores (e.g., RP2040/RP2350), this pins the
+// goroutine to the core it's currently running on.
 // Called by go1.18 standard library on windows, see https://github.com/golang/go/issues/49320
```
While here, remove this now irrelevant comment.

eliasnaur commented on src/runtime/runtime.go:
```diff
@@ -98,14 +98,18 @@ func os_sigpipe() {
 
 // LockOSThread wires the calling goroutine to its current operating system thread.
-// Stub for now
+// On microcontrollers with multiple cores (e.g., RP2040/RP2350), this pins the
+// goroutine to the core it's currently running on.
```
Is it more precise to say "with the "cores" scheduler"?

eliasnaur commented on src/machine/machine_rp2_cores.go:
```diff
@@ -0,0 +1,26 @@
+
+const numCPU = 2 // RP2040 and RP2350 both have 2 cores
+
+// LockCore implementation for the cores scheduler.
```
This needs a more detailed description. For example, it doesn't say what happens if the target core is busy. I believe `LockCore` returns. If so, this is surprising to me; I would expect that once `LockCore` returns, the calling goroutine is running on the target core.

eliasnaur reviewed (commented):
Thanks. I've commented on the implementation, but I'm still not a fan of the `LockCore` API, because it may block indefinitely if a long-running goroutine is running on the target core. In a sense, `LockCore` acts as a per-core mutex that some arbitrary other goroutine may have taken, with the usual deadlock risks.

One way of getting around this issue is by requiring `LockCore` to be called before any other goroutine has started. A good place would be in an `init` function.

amken3d committed:
Improved LockCore and UnlockCore documentation for clarity on usage, behavior, and limitations with the "cores" scheduler. Updated LockOSThread and UnlockOSThread comments to reflect core pinning behavior on RP2040/RP2350.

eliasnaur commented on src/machine/machine_rp2_cores.go:
```diff
@@ -4,15 +4,46 @@ package machine
+//
+// Valid core values are 0 and 1. Panics if core is out of range.
+//
+// Only available on RP2040 and RP2350 with the "cores" scheduler.
```
Superfluous comment. The function is only available on rpxxxx by build tags.

eliasnaur commented on src/machine/machine_rp2_cores.go:
```diff
@@ -4,15 +4,46 @@ package machine
+// After calling UnlockCore, the scheduler is free to schedule the goroutine on
+// any core for automatic load balancing.
+//
+// Only available on RP2040 and RP2350 with the "cores" scheduler.
```
Superfluous comment.

eliasnaur commented on src/machine/machine_rp2_cores.go:
```diff
@@ -4,15 +4,46 @@ package machine
+// that point, the goroutine will wait in the target core's queue if that core
+// is busy running another goroutine.
+//
+// To avoid potential blocking on a busy core, consider calling LockCore in an
+// init function before any other goroutines have started. This guarantees the
+// target core is available.
```
I think this should be a hard requirement. That is, `LockCore` should panic if any other goroutine has started.

amken3d replied:
Regarding "panic if goroutines started" - I have a specific use case for dynamic pinning:

Motion control board where:

Core 0: Communications and non-critical tasks
Core 1: Hard real-time step generation (must not be interrupted)
The pattern I need is:

func main() {
    go func() {
        machine.LockCore(1)  // Pin worker to core 1
        stepGenerationLoop()
    }()
    machine.LockCore(0)      // Pin main to core 0
    commsLoop()
}
If LockCore panics when goroutines have started, this pattern wouldn't work.

Would you accept one of these:

- Adding Gosched() in LockCore (so it actually migrates before returning)
- Keeping dynamic pinning allowed, with clear documentation of risks
- Perhaps a convention that each core should only have ONE pinned goroutine?

The deadlock risk is manageable if users follow the pattern of pinning early and using one goroutine per core for pinned work.

eliasnaur replied:
> Adding Gosched() in LockCore (so it actually migrates before returning)

Yes, if we do this `LockCore` should not return before the core is pinned.

However, given your use case: is it possible to run the step generation off a hardware timer and an interrupt handler? That seems a better fit, and you can keep both cores running non-critical code.

amken3d replied:
No, That is precisely the implementation that I am trying to get away from.

eliasnaur replied:
Can you elaborate why? Assuming your `stepGenerationLoop` sometimes sleeps, it seems much less efficient to lock a core 100% of the time, even during sleeps than running an interrupt off a timer.

amken3d replied:
The interrupt-based approach has several issues for precision step generation:
1. Interrupt Latency & Jitter Interrupts can be delayed by:
Other higher-priority interrupts
Critical sections in the runtime (GC, scheduler locks)
interrupt.Disable() calls in drivers
For step generation, even microseconds of jitter causes visible artifacts (rough surfaces, inaccurate positioning). A dedicated core can be made to have deterministic timin

2. My step generation loop doesn't time.Sleep() - it does tight timing loops or busy-waits on hardware timers for sub-microsecond precision:

func stepGenerationLoop() {
    for {
        waitForNextStepTime()  // Busy-wait or timer poll
        pulseStepPins()        // Must happen at EXACT time
        calculateNextStep()    // Acceleration/deceleration curves
    }
}

The core isn't idle - it's maintaining precise timing. An interrupt would add latency between "timer fired" and "step pin toggled."

3. Interrupt Context Limitations In an interrupt handler, I can't:
Use blocking operations
Allocate memory (GC issues)
Have complex state machines easily
Use goroutine synchronization primitives
With a dedicated core, I can write normal Go code with loops, state, and even coordinate multiple axes cleanly.

4. The Alternative is Worse Without core pinning, my options are:
Interrupt-based (jitter issues above)
PIO-only (limited for complex multi-axis coordination)
External step generator chip (added cost/complexity)
Core pinning gives me a way to isolate real-time work.

eliasnaur replied:
I see. I believe I'm doing something similar on a PIO-capable chip (rp2350). My solution is a 3-layer architecture:

- Regular Go code generates high-level primitives (line segments, bezier curves etc.) and sends them over a channel. This code is only timing sensitive to the extent that it needs to run often enough to not starve the interrupt handler. Depending on your application, you can get rid of the timing requirement by generating batches of primitives where each batch ends with the machine at a safe point (zero velocity/acceleration etc.).
- Interrupt handler receives from the channel and chops the primitives into fixed-duration updates and packs them into DMA buffers. For example, 2 steppers is 4 bits (2xdir, 2xstep) per update, and 8 updates per 32-bit PIO FIFO word.
The interrupt handler is only timing sensitive insofar it must run often enough to fill the DMA buffers. If you don't have too much going on, you may not even need DMA and can get away with feeding the PIO FIFO buffer (8 words I believe) directly.
- A PIO state machine receives updates from its FIFO (fed directly or through DMA) and multiplex them onto the corresponding GPIO pins. The nature of PIOs makes the timing very robust and decoupled from the activity of the system, except for contention on the memory bus.

amken3d replied:
There is always multiple ways to skin a cat. I am using an approach similar to what you described as well. But having core pinning has its advantages as well. Also, It seems like several people have asked for it previously. Appreciate your reviews and responses

maks replied:
As one of the other people that have asked for pinning to cores, I'll chime in with my use case: I want to be able to dedicate a RP2040 (or 2350) core to realtime audio sample generation for a [groovebox project](https://github.com/xiphonics/picoTracker). We are currently able to do this without issues using the RPI C SDK but the lack of core pinning prevents any attempt at trying to use tinyGo in the future for this or similar applications.

eliasnaur replied:
@maks your use case needs `LockOSThread` (lock to any core), not the more specific `LockCore`, right?

maks replied:
@eliasnaur sorry for slow reply. I'm new to Go so I'm unsure of how LockOSThread() is supposed to work for this context, but what I was after was to be able to run a Go routine on a specific core of a RP2040/2350 (eg. core1) and then have all other routines including main run on core0 only.

eliasnaur commented on src/machine/machine_rp2_cores.go:
```diff
@@ -4,15 +4,46 @@ package machine
+// goroutine to the target core. The actual migration happens at the next
+// scheduling point (e.g., channel operation, time.Sleep, or Gosched). After
+// that point, the goroutine will wait in the target core's queue if that core
+// is busy running another goroutine.
```
Exposing this implementation detail seems like an unnecessary burden on the caller. Why can't `LockCore` switch the goroutine over to the target core before returning? For instance, why can't `LockCore` call `Gosched`?

eliasnaur reviewed (commented):
@aykevl should probably take a look.

amken3d committed:
Add runtime.Gosched call in LockCore for improved scheduling

amken3d committed:
Removed superfluous comments

amken3d committed:
Added GoSched() to scheduler_cores.go instead of incorrectly calling it in machine_rp2_cores.go

amken3d committed:
Remove unused import of "runtime" in machine_rp2_cores.go

rdon-key commented:
I think explicit core pinning is a valid use case here, rather than merely a less efficient substitute for interrupts.

Moving the work to a timer interrupt does not remove the concurrency problem. It changes the design into communication between normal Go code and interrupt context, potentially involving shared state, queues, DMA buffers, or PIO. Each of those boundaries requires synchronization and clear ownership rules to avoid races, missed updates, buffer underruns, and latency problems.

A goroutine pinned to a dedicated core offers a substantially simpler execution model. Time-critical state can have a single owner and the processing can be written as ordinary sequential Go code. This can significantly reduce the concurrency surface compared with an interrupt-based design.

Of course, dedicating a core has a cost, and interrupts, DMA, or PIO may be more efficient for some applications. But that is an application-level tradeoff. One approach should not be rejected merely because another, substantially more complex approach is technically possible.

The relevant question is not only whether the task can be implemented using interrupts. It is whether it can be implemented with comparable simplicity, portability, and race safety. In many cases, it cannot.

This is also why `runtime.LockOSThread` is not always sufficient. It pins a goroutine to whichever core it currently happens to be running on, but it does not allow the application to select a particular core. A valid use case is to leave the runtime, USB, networking, or other existing work on core 0, while explicitly assigning time-critical work to core 1.

On the blocking concern: I agree that `LockCore` should not return until the goroutine is actually running on the requested core. The implementation should therefore include the `Gosched()`-based migration rather than merely recording the requested affinity.

The call can block indefinitely if the target core is already occupied by a pinned goroutine that never yields or blocks. However, I think this is an inherent constraint of an explicit low-level core-affinity API, and it should be documented as such.

I would prefer documenting conventions such as pinning early where possible, avoiding migration onto an already dedicated core, and avoiding pinned goroutines that monopolize a core without yielding or blocking, rather than introducing a hard “panic once other goroutines have started” rule.

Such a rule would prevent the dynamic pinning use case described by amken3d, which is one of the main reasons to provide `LockCore` in addition to `runtime.LockOSThread`.
